src/analysis/processing/qgsalgorithmdetectdatasetchanges.cpp

```diff
@@ -0,0 +1,432 @@
+    {
+      if ( originalNullGeometryAttributes.contains( attrs ) )
+      {
+        feedback->reportError( QObject::tr( "A non-unique set of comparison attributes was found for "
+                                            "one or more features without geometries - results may be misleading (features %1 and %2)" ).arg( f.id() ).arg( originalNullGeometryAttributes.value( attrs ) ) );
+      }
+      else
+      {
+        originalNullGeometryAttributes.insert( attrs, f.id() );
+      }
+    }
+
+    // overall this loop takes about 10% of time
+    current++;
+    feedback->setProgress( 0.10 * current * step );
+    return true;
+  } );
+
+  QSet<QgsFeatureId> unchangedOriginalIds;
+  QSet<QgsFeatureId> addedRevisedIds;
+  current = 0;
+
+  // second iteration: we loop through ALL revised features, checking whether each is a match for a geometry from the
+  // original set. If so, check if the feature is unchanged. If there's no match with the original features, we mark it as an "added" feature
+  step = mRevised->featureCount() > 0 ? 100.0 / mRevised->featureCount() : 0;
+  QgsFeatureRequest revisedRequest = QgsFeatureRequest().setDestinationCrs( mOriginal->sourceCrs(), context.transformContext() );
+  revisedRequest.setSubsetOfAttributes( mRevisedFieldsToCompareIndices );
+  it = mRevised->getFeatures( revisedRequest );
+  QgsFeature revisedFeature;
+  while ( it.nextFeature( revisedFeature ) )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    int idx = 0;
+    for ( int field : mRevisedFieldsToCompareIndices )
+    {
+      attrs[idx++] = revisedFeature.attributes().at( field );
+    }
+
+    bool matched = false;
+
+    if ( !revisedFeature.hasGeometry() )
+    {
+      if ( originalNullGeometryAttributes.contains( attrs ) )
+      {
+        // found a match for feature
+        unchangedOriginalIds.insert( originalNullGeometryAttributes.value( attrs ) );
+        matched = true;
+      }
+    }
+    else
+    {
+      // can we match this feature?
+      const QList<QgsFeatureId> candidates = index.intersects( revisedFeature.geometry().boundingBox() );
+
+      // lazy evaluate -- there may be NO candidates!
+      QgsGeometry revised;
+
+      for ( const QgsFeatureId candidateId : candidates )
+      {
+        if ( unchangedOriginalIds.contains( candidateId ) )
+        {
+          // already matched this original feature
+          continue;
+        }
+
+        // attribute comparison is faster to do first, if desired
+        if ( !mFieldsToCompare.empty() )
+        {
+          if ( attrs != originalAttributes[ candidateId ] )
+          {
+            // attributes don't match, so candidates is not a match
+            continue;
+          }
+        }
+
+        QgsGeometry original = originalGeometries.value( candidateId );
+        // lazy evaluation
+        if ( revised.isNull() )
+        {
+          revised = revisedFeature.geometry();
+          // drop z/m if not wanted for match
```

Comment:
Do we really want to unconditionally drop z here? I.e. is a topological match only based on a 2D representation? Wondering if it'd be better to have Topological2D and Topological3D. Or that might simply be overdoing it :)

Reply:
That would depend on support in geos. Currently it doesn't handle these, so we drop them (and this limitation is mentioned in the algorithm documentation)

Reply:
@nyalldawson , okeydokey